hecat/processors/github_metadata.py

```diff
@@ -76,41 +67,183 @@ def write_software_yaml(step, software):
-                github_url = software['source_code_url']
-        elif 'website_url' in software:
-            if re.search(r'^https://github.com/[\w\.\-]+/[\w\.\-]+/?$', software['website_url']):
-                github_url = software['website_url']
-        if github_url:
-            logging.debug('%s is a github project URL', github_url)
-            if 'gh_metadata_only_missing' in step['module_options'].keys() and step['module_options']['gh_metadata_only_missing']:
-                if ('stargazers_count' not in software) or ('updated_at' not in software) or ('archived' not in software):
-                    logging.info('Missing metadata for %s, gathering it from Github API', software['name'])
-                    gh_metadata, latest_commit_date = get_gh_metadata(step, github_url, g, errors)
-                    software['stargazers_count'] = int(gh_metadata.stargazers_count)
-                    software['updated_at'] = datetime.strftime(latest_commit_date, "%Y-%m-%d")
-                    software['archived'] = bool(gh_metadata.archived)
-                    write_software_yaml(step, software)
+                # Check if we only want to update missing metadata or all metadata
+                if 'gh_metadata_only_missing' in step['module_options'].keys() and step['module_options']['gh_metadata_only_missing']:
+                    required_keys = ['stargazers_count', 'updated_at', 'archived', 'current_release', 'last_commit_date', 'commit_history']
+                    if any(key not in software for key in required_keys):
+                        github_projects.append(software)
+                    else:
+                        logging.debug('all metadata already present, skipping %s', software['source_code_url'])
+                # If key is not present, update all metadata
                 else:
-                    logging.debug('all metadata already present, skipping %s', github_url)
+                    github_projects.append(software)
+    # Get the metadata for the GitHub repositories
+    GITHUB_GRAPHQL_API = "https://api.github.com/graphql"
+    headers = {"Authorization": f"Bearer {GITHUB_TOKEN}"}
+
+    # Get the URLs of the queued repositories
+    github_urls = [software['source_code_url'] for software in github_projects]
+    repos = [re.sub('https://github.com/', '', url) for url in github_urls]
+
+    # This limit was tested with a personal access token, batch_size = 30, timeout = 60 (new default if not provided in a config); Worked fine for the full repo (enable debug messages to see usage of API Rate limit stats)
+    # Split repo list into batches of batch_size
+    if 'batch_size' in step['module_options']:
+        batch_size = step['module_options']['batch_size']
+    else:
+        # Default batch_size of repos if not specified in config file
+        batch_size = 30
+    batches = [repos[i * batch_size:(i + 1) * batch_size] for i in range((len(repos) + batch_size - 1) // batch_size )]
```

Comment:
```suggestion
    # build a list of lists (batches) of repo names, for each batch a graphql query will be made
    batches = [repos[i * batch_size:(i + 1) * batch_size] for i in range((len(repos) + batch_size - 1) // batch_size )]
```
a little comment to clarify what it does/discussed before